### ISSUE TITLE
Unable to parse when using wasm exceptions

### ISSUE THREAD

**Issue by bkotsopoulossc** [GoogleChromeLabs/wasmbin#7]
I tried parsing a binary that uses [wasm exceptions](https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md) (`-fwasm-exceptions`) now that they seem to have good browser support ([link](https://webassembly.org/roadmap/)).

I ended up getting the error `could not recognise discriminant 0xD for type Section`

If I understand correctly, it's because we need to parse the new [tag section](https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md#tag-section) (value=13).

I assume we need to define the new section [here](https://github.com/GoogleChromeLabs/wasmbin/blob/ddb13cee611653a2840bd78aaba98b3d77ad0399/src/sections.rs#L418), and probably a few other places.

**Comment by RReverser** [comment]
Yeah that sounds about right. I can't work on it right now, but if you want to send a PR, that would be welcome. The design of the parser should make them pretty easy to add following existing examples and the binary format spec of the proposal.

**Comment by bkotsopoulossc** [comment]
I'm not that comfortable in rust, but I'm willing to take a shot! Are there any other parts of the codebase I should keep an eye out for, or should it be contained in sections.rs and mostly follow from the other examples?

**Comment by RReverser** [comment]
Ideally you should only extend the existing types, and the derive macros will take care of the rest, so you won't even need to write any custom code. Do let me know though if you run into any problems!

**Comment by bkotsopoulossc** [comment]
Sounds good. I also noticed that the test suite has a nice example of the tag section [here](https://github.com/WebAssembly/testsuite/blob/main/proposals/exception-handling/tag.wast), but this file was just recently added. If I try and update the test submodule, it looks like there are some test cases failing:
```
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/spec.rs (target/debug/deps/spec-1760614cc5dae51d)
Error: elem segment item doesn't match elem segment type
     --> tests/testsuite/elem.wast:465:49
      |
  465 |     (elem (i32.const 0) funcref (ref.null extern))
      |                                                 ^
error: test failed, to rerun pass '--test spec'
```

(Also the "simd" proposal tests have been moved into the regular test folder.)

It looks like this error is potentially coming from the `wast` crate, which is a couple versions behind [here](https://crates.io/crates/wast), but updating that brings some new compile errors in the test suite, like
```
^^^^^^ expected struct `wast::Module`, found enum `wast::QuoteModule`
```

**Comment by matusfedorko** [comment]
Is there any chance that pull requests related to this issue will get accepted anytime soon ? Yesterday I ran into this exact issue with Sentry's `wasm-split`. Would be awesome if this worked, we could finally have properly symbolicated crash reports from wasm clients.

**Comment by RReverser** [comment]
Of course they would be. Looking at the open PR, seems that after I left some comments, author stopped responding, but I'd be happy if someone finished it.

**Comment by RReverser** [comment]
Seems like a great opportunity to contribute. This parser is intentionally designed to require as little actual Rust code as possible, you just declare data structures by copy-pasting and looking up the spec and it auto-generates the rest.

You can see the PR itself that pretty much added everything needed, only had some nits left. I can fix them myself, but if you want to, could be a good first start. https://github.com/GoogleChromeLabs/wasmbin/pull/9/files

**Comment by bkotsopoulossc** [comment]
Sorry about abandoning this, I only needed it for a quick experiment and I ended up just building wasm-split against this commit and using that for what I needed.

Let's get clear on what's left here:
- naming nit
- feature flag
- upgrading the test suite to be able to turn on exceptions tests - there were challenges around that before, maybe fixed now?
- there was also discussion in #11 to combine that with #9 and merge together

**Comment by RReverser** [comment]
No problem. Sounds like @bkotsopoulossc might finish it anyway.

**Comment by RReverser** [comment]
@bkotsopoulossc Hey, so do you think you can still pick it up?

**Comment by RReverser** [comment]
Ok I merged #11 into #9, rebased onto latest cleaned-up master, and fixed up everything on the list above myself. Going to do a bit more cleanup on master and release in couple of minutes.

**Comment by RReverser** [comment]
Released in 0.4.0 now.

**Comment by RReverser** [comment]
@bkotsopoulossc Thanks for the original PR btw. Aside from cleaning up those nits, I didn't have to do anything and the testsuite - which indeed was finally updated since - ran successfully out of the box.
